File: splp-bun/src/lib/request-reply/messaging-client.ts
import { KafkaWrapper } from '../kafka/kafka-wrapper.js';
import { CassandraLogger } from '../logging/cassandra-logger.js';
import { encryptPayload, decryptPayload } from '../crypto/encryption.js';
import { generateRequestId } from '../utils/request-id.js';
import type {
  MessagingConfig,
  RequestHandler,
  HandlerRegistry,
  RequestMessage,
  ResponseMessage,
  EncryptedMessage,
} from '../../types/index.js';

export class MessagingClient {
  private kafka: KafkaWrapper;
  private logger: CassandraLogger;
  private encryptionKey: string;
  private handlers: HandlerRegistry = {};
  private pendingRequests: Map<
    string,
    {
      resolve: (value: any) => void;
      reject: (error: Error) => void;
      startTime: number;
    }
  > = new Map();

  constructor(private config: MessagingConfig) {
    this.kafka = new KafkaWrapper(config.kafka);
    this.logger = new CassandraLogger(config.cassandra);
    this.encryptionKey = config.encryption.encryptionKey;
  }

  /**
   * Initialize the messaging client - single line setup for users
   */
  async initialize(): Promise<void> {
    await this.kafka.connectProducer();
    await this.kafka.connectConsumer();
    await this.logger.initialize();
    console.log('Messaging client initialized successfully');
  }

  /**
   * Register a handler for a specific topic
   * Users only need to register their handler function
   */
  registerHandler<TRequest = any, TResponse = any>(
    topic: string,
    handler: RequestHandler<TRequest, TResponse>
  ): void {
    this.handlers[topic] = handler;
    console.log(`Handler registered for topic: ${topic}`);
  }

  /**
   * Start consuming messages and processing with registered handlers
   */
  async startConsuming(topics: string[]): Promise<void> {
    // Subscribe to request topics
    const requestTopics = topics;

    // Also subscribe to reply topics for request-reply pattern
    const replyTopic = `${this.config.kafka.clientId}.replies`;
    const allTopics = [...requestTopics, replyTopic];

    await this.kafka.subscribe(allTopics, async ({ topic, partition, message }) => {
      try {
        if (topic === replyTopic) {
          // Handle reply message
          await this.handleReply(message.value?.toString() || '');
        } else {
          // Handle request message
          await this.handleRequest(topic, message.value?.toString() || '');
        }
      } catch (error) {
        console.error('Error processing message:', error);
      }
    });

    console.log(`Started consuming from topics: ${allTopics.join(', ')}`);
  }

  /**
   * Send a request with automatic encryption and wait for reply
   * Returns the decrypted response
   */
  async request<TRequest = any, TResponse = any>(
    topic: string,
    payload: TRequest,
    timeoutMs: number = 30000
  ): Promise<TResponse> {
    const requestId = generateRequestId();
    const startTime = Date.now();

    // Create promise for reply
    const replyPromise = new Promise<TResponse>((resolve, reject) => {
      this.pendingRequests.set(requestId, {
        resolve,
        reject,
        startTime,
      });

      // Set timeout
      setTimeout(() => {
        if (this.pendingRequests.has(requestId)) {
          this.pendingRequests.delete(requestId);
          reject(new Error(`Request timeout after ${timeoutMs}ms`));
        }
      }, timeoutMs);
    });

    // Encrypt payload
    const encryptedMessage = encryptPayload(payload, this.encryptionKey, requestId);

    // Log request
    await this.logger.log({
      request_id: requestId,
      timestamp: new Date(),
      type: 'request',
      topic,
      payload,
    });

    // Send to Kafka
    const replyTopic = `${this.config.kafka.clientId}.replies`;
    const messageWithReplyTopic = {
      ...encryptedMessage,
      replyTopic,
    };

    await this.kafka.sendMessage(
      topic,
      JSON.stringify(messageWithReplyTopic),
      requestId
    );

    console.log(`Request sent: ${requestId} to topic: ${topic}`);

    return replyPromise;
  }

  /**
   * Handle incoming request - decrypt, process with handler, encrypt and send reply
   */
  private async handleRequest(topic: string, messageValue: string): Promise<void> {
    const startTime = Date.now();
    let requestId = '';

    try {
      // Parse message
      const encryptedMessage: EncryptedMessage & { replyTopic?: string } = JSON.parse(messageValue);
      requestId = encryptedMessage.request_id;

      // Decrypt payload
      const { payload } = decryptPayload(encryptedMessage, this.encryptionKey);

      console.log(`Request received: ${requestId} from topic: ${topic}`);

      // Find handler
      const handler = this.handlers[topic];
      if (!handler) {
        throw new Error(`No handler registered for topic: ${topic}`);
      }

      // Process with handler
      const responsePayload = await handler(requestId, payload);

      // Create response
      const response: ResponseMessage = {
        request_id: requestId,
        payload: responsePayload,
        timestamp: Date.now(),
        success: true,
      };

      // Encrypt response
      const encryptedResponse = encryptPayload(response, this.encryptionKey, requestId);

      // Send reply to reply topic
      const replyTopic = encryptedMessage.replyTopic || `${topic}.replies`;
      await this.kafka.sendMessage(
        replyTopic,
        JSON.stringify(encryptedResponse),
        requestId
      );

      // Log response
      const duration = Date.now() - startTime;
      await this.logger.log({
        request_id: requestId,
        timestamp: new Date(),
        type: 'response',
        topic,
        payload: responsePayload,
        success: true,
        duration_ms: duration,
      });

      console.log(`Response sent: ${requestId} (${duration}ms)`);
    } catch (error) {
      // Log error
      const duration = Date.now() - startTime;
      await this.logger.log({
        request_id: requestId,
        timestamp: new Date(),
        type: 'response',
        topic,
        payload: null,
        success: false,
        error: error instanceof Error ? error.message : String(error),
        duration_ms: duration,
      });

      console.error(`Error handling request ${requestId}:`, error);
    }
  }

  /**
   * Handle incoming reply - decrypt and resolve pending request
   */
  private async handleReply(messageValue: string): Promise<void> {
    try {
      // Parse encrypted message
      const encryptedMessage: EncryptedMessage = JSON.parse(messageValue);
      const requestId = encryptedMessage.request_id;

      // Decrypt response
      const { payload: response } = decryptPayload<ResponseMessage>(
        encryptedMessage,
        this.encryptionKey
      );

      console.log(`Reply received: ${requestId}`);

      // Find pending request
      const pendingRequest = this.pendingRequests.get(requestId);
      if (pendingRequest) {
        this.pendingRequests.delete(requestId);

        // Log response
        const duration = Date.now() - pendingRequest.startTime;
        await this.logger.log({
          request_id: requestId,
          timestamp: new Date(),
          type: 'response',
          topic: 'reply',
          payload: response.payload,
          success: response.success,
          error: response.error,
          duration_ms: duration,
        });

        // Resolve or reject promise
        if (response.success) {
          pendingRequest.resolve(response.payload);
        } else {
          pendingRequest.reject(new Error(response.error || 'Request failed'));
        }
      }
    } catch (error) {
      console.error('Error handling reply:', error);
    }
  }

  /**
   * Get logger instance for manual queries
   */
  getLogger(): CassandraLogger {
    return this.logger;
  }

  /**
   * Close all connections
   */
  async close(): Promise<void> {
    await this.kafka.disconnect();
    await this.logger.close();
    console.log('Messaging client closed');
  }
}

File: splp-java/src/main/java/com/perlinsos/splp/messaging/MessagingClient.java
package com.perlinsos.splp.messaging;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.perlinsos.splp.crypto.EncryptionService;
import com.perlinsos.splp.kafka.KafkaWrapper;
import com.perlinsos.splp.logging.CassandraLogger;
import com.perlinsos.splp.types.*;
import com.perlinsos.splp.utils.RequestIdGenerator;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.time.Instant;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.TimeUnit;

/**
 * Messaging Client for request-reply patterns with encryption and logging
 */
public class MessagingClient {
    private static final Logger logger = LoggerFactory.getLogger(MessagingClient.class);
    private static final ObjectMapper objectMapper = new ObjectMapper();

    private final MessagingConfig config;
    private final KafkaWrapper kafka;
    private final CassandraLogger cassandraLogger;
    private final String encryptionKey;
    private final HandlerRegistry handlers;
    private final Map<String, PendingRequest> pendingRequests = new ConcurrentHashMap<>();

    public MessagingClient(MessagingConfig config) {
        this.config = config;
        this.kafka = new KafkaWrapper(config.getKafkaConfig());
        this.cassandraLogger = new CassandraLogger(config.getCassandraConfig());
        this.encryptionKey = config.getEncryptionConfig().getEncryptionKey();
        this.handlers = new HandlerRegistry();
    }

    /**
     * Initialize the messaging client - single line setup for users
     */
    public CompletableFuture<Void> initialize() {
        return CompletableFuture.allOf(
            kafka.connectProducer().thenApply(p -> null),
            kafka.connectConsumer().thenApply(c -> null),
            cassandraLogger.initialize()
        ).thenRun(() -> logger.info("Messaging client initialized successfully"));
    }

    /**
     * Register a handler for a specific topic
     * Users only need to register their handler function
     */
    public <TRequest, TResponse> void registerHandler(String topic, RequestHandler<TRequest, TResponse> handler) {
        handlers.registerHandler(topic, handler);
        logger.info("Handler registered for topic: {}", topic);
    }

    /**
     * Start consuming messages and processing with registered handlers
     */
    public CompletableFuture<Void> startConsuming(List<String> topics) {
        // Subscribe to request topics
        List<String> requestTopics = topics;

        // Also subscribe to reply topics for request-reply pattern
        String replyTopic = config.getKafkaConfig().getClientId() + ".replies";
        List<String> allTopics = new java.util.ArrayList<>(requestTopics);
        allTopics.add(replyTopic);

        return kafka.subscribe(allTopics, payload -> {
            try {
                if (replyTopic.equals(payload.getTopic())) {
                    // Handle reply message
                    handleReply(payload.getValue()).join();
                } else {
                    // Handle request message
                    handleRequest(payload.getTopic(), payload.getValue()).join();
                }
            } catch (Exception e) {
                logger.error("Error processing message", e);
            }
        }).thenRun(() -> logger.info("Started consuming from topics: {}", String.join(", ", allTopics)));
    }

    /**
     * Send a request with automatic encryption and wait for reply
     * Returns the decrypted response
     */
    public <TRequest, TResponse> CompletableFuture<TResponse> request(String topic, TRequest payload, long timeoutMs) {
        String requestId = RequestIdGenerator.generate();
        long startTime = System.currentTimeMillis();

        // Create promise for reply
        CompletableFuture<TResponse> replyFuture = new CompletableFuture<>();
        
        PendingRequest pendingRequest = new PendingRequest(replyFuture, startTime);
        pendingRequests.put(requestId, pendingRequest);

        // Set timeout
        CompletableFuture.delayedExecutor(timeoutMs, TimeUnit.MILLISECONDS).execute(() -> {
            PendingRequest pending = pendingRequests.remove(requestId);
            if (pending != null) {
                pending.future.completeExceptionally(
                    new RuntimeException("Request timeout after " + timeoutMs + "ms")
                );
            }
        });

        return CompletableFuture.runAsync(() -> {
            try {
                // Encrypt payload
                EncryptionService.EncryptionResult encryptedMessage = EncryptionService.encryptPayload(payload, encryptionKey, requestId);

                // Log request
                LogEntry requestLog = LogEntry.createRequestLog(requestId, Instant.now(), topic, payload);
                cassandraLogger.log(requestLog).join();

                // Send to Kafka
                String replyTopicName = config.getKafkaConfig().getClientId() + ".replies";
                MessageWithReplyTopic messageWithReplyTopic = new MessageWithReplyTopic(
                    encryptedMessage.getRequestId(),
                    encryptedMessage.getData(),
                    encryptedMessage.getIv(),
                    encryptedMessage.getTag(),
                    replyTopicName
                );

                String messageJson = objectMapper.writeValueAsString(messageWithReplyTopic);
                kafka.sendMessage(topic, messageJson, requestId).join();

                logger.info("Request sent: {} to topic: {}", requestId, topic);
            } catch (Exception e) {
                PendingRequest pending = pendingRequests.remove(requestId);
                if (pending != null) {
                    pending.future.completeExceptionally(e);
                }
                throw new RuntimeException("Failed to send request", e);
            }
        }).thenCompose(v -> replyFuture);
    }

    /**
     * Send a request with default timeout (30 seconds)
     */
    public <TRequest, TResponse> CompletableFuture<TResponse> request(String topic, TRequest payload) {
        return request(topic, payload, 30000);
    }

    /**
     * Handle incoming request - decrypt, process with handler, encrypt and send reply
     */
    private CompletableFuture<Void> handleRequest(String topic, String messageValue) {
        return CompletableFuture.runAsync(() -> {
            long startTime = System.currentTimeMillis();
            String requestId = "";

            try {
                // Parse message
                MessageWithReplyTopic encryptedMessage = objectMapper.readValue(messageValue, MessageWithReplyTopic.class);
                requestId = encryptedMessage.getRequestId();

                // Decrypt payload
                var decryptionResult = EncryptionService.decryptPayload(
                    new EncryptedMessage(encryptedMessage.getRequestId(), encryptedMessage.getData(), 
                                       encryptedMessage.getIv(), encryptedMessage.getTag()),
                    encryptionKey,
                    Object.class
                );
                Object payload = decryptionResult.getPayload();

                logger.info("Request received: {} from topic: {}", requestId, topic);

                // Find handler
                RequestHandler<Object, Object> handler = handlers.getHandler(topic);
                if (handler == null) {
                    throw new RuntimeException("No handler registered for topic: " + topic);
                }

                // Process with handler
                CompletableFuture<Object> handlerResult = handler.handle(requestId, payload);
                Object responsePayload = handlerResult.join();

                // Create response
                ResponseMessage<Object> response = ResponseMessage.success(requestId, responsePayload, Instant.now());

                // Encrypt response
                EncryptionService.EncryptionResult encryptedResponse = EncryptionService.encryptPayload(response, encryptionKey, requestId);

                // Send reply to reply topic
                String replyTopic = encryptedMessage.getReplyTopic() != null ? 
                    encryptedMessage.getReplyTopic() : topic + ".replies";
                
                EncryptedMessage replyMessage = new EncryptedMessage(
                    encryptedResponse.getRequestId(),
                    encryptedResponse.getData(),
                    encryptedResponse.getIv(),
                    encryptedResponse.getTag()
                );

                String replyJson = objectMapper.writeValueAsString(replyMessage);
                kafka.sendMessage(replyTopic, replyJson, requestId).join();

                // Log response
                long duration = System.currentTimeMillis() - startTime;
                LogEntry responseLog = LogEntry.createResponseLog(requestId, Instant.now(), topic, responsePayload, true, null, (int) duration);
                cassandraLogger.log(responseLog).join();

                logger.info("Response sent: {} ({}ms)", requestId, duration);

            } catch (Exception error) {
                // Log error
                long duration = System.currentTimeMillis() - startTime;
                LogEntry errorLog = LogEntry.createErrorLog(requestId, Instant.now(), topic, null, error.getMessage(), (int) duration);
                cassandraLogger.log(errorLog).join();

                logger.error("Error handling request {}: {}", requestId, error.getMessage(), error);
            }
        });
    }

    /**
     * Handle incoming reply - decrypt and resolve pending request
     */
    private CompletableFuture<Void> handleReply(String messageValue) {
        return CompletableFuture.runAsync(() -> {
            try {
                // Parse encrypted message
                EncryptedMessage encryptedMessage = objectMapper.readValue(messageValue, EncryptedMessage.class);
                String requestId = encryptedMessage.getRequestId();

                // Decrypt response
                var decryptionResult = EncryptionService.decryptPayload(encryptedMessage, encryptionKey, ResponseMessage.class);
                ResponseMessage<?> response = decryptionResult.getPayload();

                logger.info("Reply received: {}", requestId);

                // Find pending request
                PendingRequest pendingRequest = pendingRequests.remove(requestId);
                if (pendingRequest != null) {
                    // Log response
                    long duration = System.currentTimeMillis() - pendingRequest.startTime;
                    LogEntry responseLog = LogEntry.createResponseLog(requestId, Instant.now(), "reply", 
                        response.getPayload(), response.getSuccess(), response.getError(), (int) duration);
                    cassandraLogger.log(responseLog).join();

                    // Resolve or reject promise
                    if (response.getSuccess()) {
                        pendingRequest.future.complete(response.getPayload());
                    } else {
                        pendingRequest.future.completeExceptionally(
                            new RuntimeException(response.getError() != null ? response.getError() : "Request failed")
                        );
                    }
                }
            } catch (Exception error) {
                logger.error("Error handling reply", error);
            }
        });
    }

    /**
     * Get logger instance for manual queries
     */
    public CassandraLogger getLogger() {
        return cassandraLogger;
    }

    /**
     * Close all connections
     */
    public CompletableFuture<Void> close() {
        return CompletableFuture.allOf(
            kafka.disconnect(),
            cassandraLogger.close()
        ).thenRun(() -> logger.info("Messaging client closed"));
    }

    /**
     * Internal class for tracking pending requests
     */
    private static class PendingRequest {
        final CompletableFuture<Object> future;
        final long startTime;

        PendingRequest(CompletableFuture<?> future, long startTime) {
            this.future = (CompletableFuture<Object>) future;
            this.startTime = startTime;
        }
    }

    /**
     * Internal class for messages with reply topic
     */
    private static class MessageWithReplyTopic extends EncryptedMessage {
        private String replyTopic;

        public MessageWithReplyTopic() {
            super();
        }

        public MessageWithReplyTopic(String requestId, String data, String iv, String tag, String replyTopic) {
            super(requestId, data, iv, tag);
            this.replyTopic = replyTopic;
        }

        public String getReplyTopic() {
            return replyTopic;
        }

        public void setReplyTopic(String replyTopic) {
            this.replyTopic = replyTopic;
        }
    }
}